_posts/2018-10-08-http-part5.markdown

```diff
@@ -0,0 +1,305 @@
+---
+layout: post
+title:  "The HTTP part 5 - Persistent connections"
+date:   2018-10-07 15:10:31 +0530
+categories: ["programming", "http", "server", "kotlin"]
+comments: true
+---
+## Persistent connections
+Our current flow to receive a response from server looks like this: open a connection, read a request, prepare the response, write the response, close the connection. This is suboptimal! If a client wants to send 50 requests, it has to open and close connection for each request. Establishing a connection is a costly operation, the obvious optimisation is to reuse a connection between multiple requests.
+
+![persistent connection]({{ "/assets/http5/persistent_connection.png" | absolute_url }})
+*source: [Wikipedia](https://en.wikipedia.org/wiki/HTTP_persistent_connection)*
+
+The left schema shows the current flow, the right one is what we want to achieve.
+
+Let’s go to [reddit.com](http://reddit.com) and see how many requests are sent.
+
+[![reddit connections]({{ "/assets/http5/reddit_connections.png" | absolute_url }})]({{ "/assets/http5/reddit_connections.png" | absolute_url }})
+
+The browser sent 142 requests. What's more, when you enter a website you probably start going to subpages, clicking here and there. For one website visit you can send hundreds or thousands of requests. Imagine opening and closing connection for each one of them.
+
+### Multiple persistent connections
+A browser could open one connection and download every file one by one, but this way it wouldn’t utilise all the resources available.
+Instead of this, the browser opens a few connections to a server, to load multiple files at once. HTTP/2.0 solves this problem, by enabling the client to send multiple concurrent requests over one connection. 
+
+### Connection: keep-alive
+There is no information in [RFC1945](https://tools.ietf.org/html/rfc1945) (HTTP/1.0) about persistent connections nor `Connection` header. It was added to the protocol later on in [RFC2068](https://tools.ietf.org/html/rfc2068#section-19.7.1.1) (HTTP/1.1).
+
+You send a request to a server with a header `Connection: keep-alive`, the server will keep that connection open for some time and send the same header in the response. Now the client will also keep the connection open for some time so they can exchange data on a single connection.
+
+It was such an important optimisation that all connections [are persistent by default](https://tools.ietf.org/html/rfc2068#section-8.1.2) unless specified otherwise. You don’t have to send a `Connection: keep-alive` header, but if you want the connection to be closed you should send a `Connection: close`.
+
+### How long to keep a connection open?
+We have to close a connection at some point. How long should we wait since the last request arrived? If we wait too long, we may end up with a bunch of useless connections. If we wait too short, clients may end up not sending multiple requests on one connection.
```

Comment:
At some point we have to close a connection
or
Eventually we have to close a connection